### PR TITLE
fix for #1611

### LINKED ISSUE #1611: setter/getter of Proxy class uses wrong field index if field index differs in multiple Realms

### PR DESCRIPTION
**Pull request by zaki50**
Problem:
  The Proxy class of each model  holds column indices in static fields.
  This works only when all Realm databases have the same column index in every column of the model class.
  If Realm databases have different column index, the getter/setter of Proxy class reads/writes wrong column.

Solution:
  This commit introduces `ColumnInfo` object that holds column index information per Realm instance which shares the same database file.

TODO:
- [x] implementation
- [x] changelog
- [x] fix failed tests for annotation processor
- [x] add test case to reproduce #1611

Fixes #1611

### CODE REVIEW

**Comment by zaki50** [comment]
retest this please

File: changelog.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@
  * Fixed a bug in RealmQuery.isNull() and RealmQuery.isNotNull() which validated the query prematurely.
  * Fixed a bug where closed Realms were trying to refresh themselves resulting in a NullPointerException.
  * Fixed a bug that made it possible to migrate open Realms, which could cause undefined behavior when querying, reading or writing data.
+ * Fixed a bug that RealmObject accesses wrong columns in some corner cases.
```
Comment (nhachicha) [inline changelog.txt]: maybe rephrase to:
Fixed a bug causing columns indices to be wrong for some edge cases

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 2]
```diff
@@ -130,10 +131,44 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+            writer.emitField("long", columnIndexVarName(variableElement),
+                    EnumSet.of(Modifier.PUBLIC, Modifier.FINAL));
+            constructorParams.add("long");
+            constructorParams.add(columnIndexVarName(variableElement));
+        }
+        writer.emitEmptyLine();
+
+        // constructor
+        writer.beginConstructor(EnumSet.noneOf(Modifier.class), constructorParams, null);
+        writer.emitStatement("super(%1$s)", "indicesMap");
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: why not just `"super(indicesMap)"`?

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 3]
```diff
@@ -585,6 +622,7 @@ private void emitCopyOrUpdateMethod(JavaWriter writer) throws IOException {
                 .beginControlFlow("if (rowIndex != TableOrView.NO_MATCH)")
                     .emitStatement("realmObject = new %s()", Utils.getProxyClassName(className))
                     .emitStatement("realmObject.realm = realm")
+                    .emitStatement("realmObject.updateColumnInfo(realm.getColumnInfo(%s.class))", className)
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: why have a `updateColumnInfo` method on `RealmObject` when it already has a reference to `realm` which has the information?

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 4]
```diff
@@ -843,6 +881,7 @@ private void emitCreateOrUpdateUsingJsonObject(JavaWriter writer) throws IOExcep
                         .beginControlFlow("if (rowIndex != TableOrView.NO_MATCH)")
                             .emitStatement("obj = new %s()", Utils.getProxyClassName(className))
                             .emitStatement("obj.realm = realm")
+                            .emitStatement("obj.updateColumnInfo(realm.getColumnInfo(%s.class))", className)
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: same

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 5]
```diff
@@ -245,15 +263,7 @@ public static void validateTable(ImplicitTransaction transaction) {
...
-            INDEX_COLUMNSTRING = table.getColumnIndex("columnString");
-            INDEX_COLUMNLONG = table.getColumnIndex("columnLong");
-            INDEX_COLUMNFLOAT = table.getColumnIndex("columnFloat");
-            INDEX_COLUMNDOUBLE = table.getColumnIndex("columnDouble");
-            INDEX_COLUMNBOOLEAN = table.getColumnIndex("columnBoolean");
-            INDEX_COLUMNDATE = table.getColumnIndex("columnDate");
-            INDEX_COLUMNBINARY = table.getColumnIndex("columnBinary");
-            INDEX_COLUMNOBJECT = table.getColumnIndex("columnObject");
-            INDEX_COLUMNREALMLIST = table.getColumnIndex("columnRealmList");
+            final AllTypesColumnInfo columnInfo = new AllTypesColumnInfo(columnIndices, table.getColumnIndex("columnString"), table.getColumnIndex("columnLong"), table.getColumnIndex("columnFloat"), table.getColumnIndex("columnDouble"), table.getColumnIndex("columnBoolean"), table.getColumnIndex("columnDate"), table.getColumnIndex("columnBinary"), table.getColumnIndex("columnObject"), table.getColumnIndex("columnRealmList"));
```
Comment (cmelchior) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Is it more performant to do this in the constructor? I preferred the old formatting as it was way more readable. The variables could just be local.

File: realm/src/main/java/io/realm/Realm.java [hunk 6]
```diff
@@ -721,9 +725,18 @@ void remove(Class<? extends RealmObject> clazz, long objectIndex) {
...
         result.row = row;
         result.realm = this;
+        result.updateColumnInfo(getColumnInfo(clazz));
         return result;
     }
 
+    ColumnInfo getColumnInfo(Class<? extends RealmObject> clazz) {
+        final ColumnInfo columnInfo = columnIndices.getColumnInfo(clazz);
+        if (columnInfo == null) {
+            throw new IllegalStateException("no validated schema information found for " + configuration.getSchemaMediator().getTableName(clazz));
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: no -> No

File: realm/src/main/java/io/realm/RealmObject.java [hunk 7]
```diff
@@ -140,6 +141,10 @@ protected static Row getRow(RealmObject obj) {
         return obj.row;
     }
 
+    protected void updateColumnInfo(ColumnInfo info) {
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmObject.java]: I'm still missing the use case? But at least remove `protected`, if it is `package private` users cannot accidentally  autocomplete into it and all our proxy classes are in `io.realm` anyway.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 8]
```diff
@@ -462,6 +462,11 @@ public int hashCode() {
     }
 
     @Override
+    protected void updateColumnInfo(ColumnInfo info) {
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: No need to override this. This method should never be called on DynamicRealmObject?

File: realm/src/main/java/io/realm/internal/ColumnIndices.java [hunk 9]
```diff
@@ -27,19 +25,16 @@
...
-     */
-    public void addClass(Class<? extends RealmObject> clazz, Map<String, Long> indices) {
-        classes.put(clazz, indices);
+    public ColumnIndices(Map<Class<? extends RealmObject>, ColumnInfo> classes) {
+        this.classes = classes;
     }
 
     /**
-     * Return mappings for the given class or null if no mapping exists.
+     * Return ColumnInfo for the given class or null if no mapping exists.
```
Comment (nhachicha) [inline realm/src/main/java/io/realm/internal/ColumnIndices.java]: javadoc tag `{@link }`

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 2 above)
Comment (zaki50) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: you are right. I'll fix this.

File: realm/src/main/java/io/realm/RealmObject.java (hunk 7 above)
Comment (nhachicha) [inline realm/src/main/java/io/realm/RealmObject.java]: you can have the method `final` & nobody will override it

Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmObject.java]: Our proxy classes has to override it :(

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 3 above)
Comment (zaki50) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: `Realm` class has following code and it requires `updateColumnInfo() on `RealmObject`.

```
    <E extends RealmObject> E get(Class<E> clazz, long rowIndex) {
        Table table = getTable(clazz);
        UncheckedRow row = table.getUncheckedRow(rowIndex);
        E result = configuration.getSchemaMediator().newInstance(clazz);
        result.row = row;
        result.realm = this;
        result.updateColumnInfo(getColumnInfo(clazz));
        return result;
    }
```

`Realm` class couldn't assign to the field of Proxy class.
If there is a way for `Realm` to assign value to Proxy class, we can remove `updateColumnInfo() from `RealmObject`.

Comment (zaki50) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: One possible way is adding args to the constructor.
I'll try that way.

Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: There is a RealmObjectProxy interface you could use instead?
Also, my brain just understood why you need to have an `updateColumnIndices` on the proxy classes, you need some way to parse information to all the accessors - DOH

I would use the RealmObjectProxy interface instead, we should avoid adding methods to RealmObject if at all possible because it impact what the user can do.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 10]
```diff
@@ -526,6 +562,14 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
         writer.emitEmptyLine();
     }
 
+    private void emitUpdateColumnInfoMethod(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod("void", "updateColumnInfo", EnumSet.of(Modifier.PROTECTED, Modifier.FINAL), "ColumnInfo", "info");
```
Comment (nhachicha) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: maybe this should be package private, remove `Modifier.PROTECTED`

**Comment by nhachicha** [comment]
is it possible to have a unit test, that reproduce #1611 maybe something where the same model is used in different Realm file/module ...?

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 3 above)
Comment (zaki50) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Oh, I already implemented the way of constructor.
If it's better to use reimplement with `RealmObjectProxy`,please comment.

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 5 above)
Comment (zaki50) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: yes, I'll fix.

**Comment by zaki50** [comment]
@nhachicha sure

**Comment by zaki50** [comment]
@nhachicha added test case to reproduce #1611

Running this test case(setter with wrong column index) against current master causes following RealmException.

```
io.realm.exceptions.RealmError: Unrecoverable error. Data type mismatch in io_realm_internal_UncheckedRow.cpp line 298
at io.realm.internal.UncheckedRow.nativeSetString(Native Method)
at io.realm.internal.UncheckedRow.setString(UncheckedRow.java:222)
at io.realm.StringAndIntRealmProxy.setStr(StringAndIntRealmProxy.java:56)
at io.realm.RealmObjectTest$5.execute(RealmObjectTest.java:880)
at io.realm.Realm.executeTransaction(Realm.java:996)
at io.realm.RealmObjectTest.testRealmProxyColumnIndex(RealmObjectTest.java:872)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:191)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:176)
at android.test.InstrumentationTestRunner.onStart(InstrumentationTestRunner.java:555)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1879)
```

getter with wrong column index causes native crash.

Both problems are fixed by this PR.

**Comment by zaki50** [comment]
please review again @realm/java

File: realm/src/main/java/io/realm/Realm.java [hunk 11]
```diff
@@ -306,18 +310,15 @@ private static Realm createAndValidate(RealmConfiguration configuration, boolean
                     throw e;
                 }
             }
-            setupColumnIndices(realm);
+            final Map<Class<? extends RealmObject>, ColumnInfo> columnInfoMap = validatedRealmFiles.get(configuration.getPath());
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: Any reason this is final? Also we could probably either remove the setupColumnIndices method or move this line down into it as well?

File: realm/src/main/java/io/realm/internal/ColumnInfo.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package io.realm.internal;
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/internal/ColumnInfo.java]: Missing license header

**Comment by cmelchior** [comment]
I need to go over this again once I am a little more awake to make sure that the columnIndices map accesses are threadsafe, but apart from that I think this looks good.

File: changelog.txt [hunk 13]
```diff
@@ -16,6 +16,7 @@
  * Fixed a bug in RealmQuery.isNull() and RealmQuery.isNotNull() which validated the query prematurely.
  * Fixed a bug where closed Realms were trying to refresh themselves resulting in a NullPointerException.
  * Fixed a bug that made it possible to migrate open Realms, which could cause undefined behavior when querying, reading or writing data.
+ * Fixed a bug causing columns indices to be wrong for some edge cases.
```
Comment (kneth) [inline changelog.txt]: "columns" -> "column"

Comment (bmunkholm) [inline changelog.txt]: For all bug descriptions, I think it's best to mention how the user would experience the bug. Referring to the issue number would be a good practice to introduce.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 14]
```diff
@@ -833,15 +853,17 @@ private void emitCreateOrUpdateUsingJsonObject(JavaWriter writer) throws IOExcep
         } else {
             String pkType = Utils.isString(metadata.getPrimaryKey()) ? "String" : "Long";
             writer
-                .emitStatement("%s obj = null", className)
+                    .emitStatement("%s obj = null", className)
```
Comment (bmunkholm) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: wrong indent?

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 15]
```diff
@@ -29,18 +27,81 @@
...
+        public final long columnLongIndex;
+        public final long columnFloatIndex;
+        public final long columnDoubleIndex;
+        public final long columnBooleanIndex;
+        public final long columnDateIndex;
+        public final long columnBinaryIndex;
+        public final long columnObjectIndex;
+        public final long columnRealmListIndex;
+
+        AllTypesColumnInfo(Table table, ImplicitTransaction transaction) {
```
Comment (bmunkholm) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Why pass the transaction as a parameter? you only need the the transaction.getPath() right? So pass that instead.

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 16]
```diff
@@ -29,18 +27,81 @@
...
+            final Map<String, Long> indicesMap = new HashMap<String, Long>(9);
+            this.columnStringIndex = table.getColumnIndex("columnString");
```
Comment (bmunkholm) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: This generates a lot of code for all fields. Can't this be reduced?

**Comment by zaki50** [comment]
@cmelchior I modified to cache entire `ColumnInfo` instead of `Map<Class<? extends RealmObject>, ColumnInfo>`. I think it does not affect thread safety. Please double check 0a2d944 .

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 14 above)
Comment (zaki50) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I'll fix this.

File: realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java [hunk 17]
```diff
@@ -26,13 +24,47 @@
 import org.json.JSONObject;
 import some.test.Booleans;
 
-public class BooleansRealmProxy extends Booleans implements RealmObjectProxy {
+public class BooleansRealmProxy extends Booleans
+        implements RealmObjectProxy {
```
Comment (bmunkholm) [inline realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java]: No need to wrap line

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 15 above)
Comment (zaki50) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: sure

File: realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java (hunk 17 above)
Comment (zaki50) [inline realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java]: This code is a correct answer to annotation processor.
So I think it's good to keep as generated.

**Comment by zaki50** [comment]
@bmunkholm sure

**Comment by zaki50** [comment]
@bmunkholm done 8803c6c

File: realm/src/androidTest/java/io/realm/RealmObjectTest.java [hunk 18]
```diff
@@ -812,4 +815,71 @@ public void testFloatPositiveInfinity() {
...
+                        final Table table = realm.getTable(StringAndInt.class);
+                        final long strIndex = table.getColumnIndex("str");
+                        final long numberIndex = table.getColumnIndex("number");
+
+                        while (0 < table.getColumnCount()) {
+                            table.removeColumn(0);
+                        }
+
+                        final long newStrIndex;
+                        final long newNumberIndex;
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmObjectTest.java]: I don't think you're using `newNumberIndex` only `newStrIndex` ?

**Comment by kneth** [comment]
:+1:

**Comment by nhachicha** [comment]
@zaki50 LGTM beside my small comments

File: realm/src/androidTest/java/io/realm/RealmObjectTest.java (hunk 18 above)
Comment (zaki50) [inline realm/src/androidTest/java/io/realm/RealmObjectTest.java]: sure. I'll remove that.

**Comment by dalinaum** [comment]
:+1:

**Comment by zaki50** [comment]
Squashed and merged.
Thank you all for lots of advises! @realm/java

**Comment by bmunkholm** [comment]
:+1: Great work @zaki50.

**Comment by dalinaum** [comment]
Great.
